### ISSUE TITLE
[TFLite] GPU delegation issues with Vision Transformer (ViT-base) in TFLite: dynamic shape & op support problems

### ISSUE THREAD

**Issue by alstjd025** [tensorflow/tensorflow#92098]
### Issue type

Support

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

source

### TensorFlow version

tf 2.20

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Mobile device

Android 13

### Python version

3.9

### Bazel version

6.1

### Current behavior?

Hi TensorFlow Lite team,

I'm currently attempting to delegate a Vision Transformer model (ViT-base, https://huggingface.co/google/vit-base-patch16-224) to the GPU using TFLite, but I’ve encountered several issues related to dynamic shapes and operator support that are blocking the GPU delegation.

When attempting GPU delegation, the process **fails during the Subgraph::PrepareOpsStartingAt() step**.
(i used max_delegate_partition to 1000 for maximum delegation)

Specifically, the Reshape operator  prepare returns **kTfLiteOutputShapeNotKnown** during its op_prepare() phase.

Structurally, ViT models have fixed input shapes (e.g., 224x224x3), so most of the downstream shapes should be statically determined.
Q1: Why does TFLite treat the output shape as unknown in this context?

I also tried modifying the delegate partitioning to only allow compatible ops for GPU delegation.
Q2: Is it possible (and safe) to manually restrict GPU delegation to only shape-safe operators to avoid this issue? What would be the best way to do that? Would there be any consistency issues across CPU-GPU boundaries?

As an experiment, I forcibly bypassed the kTfLiteOutputShapeNotKnown return (i.e., returning kTfLiteOk unconditionally), and allowed a large max_delegate_partitions (e.g., 1000). This allowed delegation to proceed further, but it failed later in the Multi-Head Attention block—specifically on the fully connected (FC) layer. Specifically, at delegation/gpu/common/model_builder.cc/    
   FullyConnectedOperationParser(), at line 1351
    if (input->tensor.shape.c != attr.weights.shape.i) {
      return absl::UnimplementedError(
          "Amount of input channels should match weights width");
    }

Q3: Why would a standard FC layer be incompatible with GPU delegation? Is this expected? Is there any workaround or support planned?

Final Question
Is there a straightforward way to make ViT models more compatible with GPU delegation in TFLite (e.g., rewriting the model, using certain converters/options)? Or would it require modifying TFLite source code (e.g., how it handles dynamic shapes or op registration)?

I’d really appreciate any advice or pointers. Thank you!

### Standalone code to reproduce the issue

```shell
just benchmark code
```

### Relevant log output

```shell
INFO: Created TensorFlow Lite delegate for GPU.
INFO: GPU delegate created.
WARNING: Attempting to use a delegate that only supports static-sized tensors with a graph that has dynamic-sized tensors (tensor#-1 is a dynamic-sized tensor).
ERROR: Failed to apply GPU delegate.
ERROR: Benchmarking failed.
--------- this is the first error from RESHAPE operator --------
```

**Comment by gaikwadrahul8** [comment]
Hi, @alstjd025 
Thank you for reporting this issue and providing detailed context. GPU delegate requires static shapes at preparation but when the GPU delegate reports dynamic-sized tensor it means that at the time of preparation it could not determine the full static shape of that tensor. This can happen even if your model's input has a fixed size due to various factors within the graph like explicit dynamic shapes during conversion using `None` or `-1` in the `input_shape` definition when converting the model or adding experimental dynamic shape converters or flags.

Certain TensorFlow operations when converted to TFLite make it hard for the converter or the delegate to statically determine the output shape even with fixed inputs in this case with `Reshape` with `-1` 

If the CPU based graph execution does satisfy your case you can consider running the graph without enabling GPU delegate.

If the graph does not have any control flow operations `tf.while_loop` (`While` op in TFLite) and `tf.cond` (`If` op in TFLite) often lead to dynamic shapes or are simply unsupported by the GPU delegate it may cause this issue so you can determine them with the [Netron](https://netron.app/) visualizer) and please refer GPU delegates supported operations for LiteRT [here](https://ai.google.dev/edge/litert/performance/gpu)

Let me answers your questions as far my current understanding
Q1: **Why unknown Reshape output shape?** Most likely, even with a fixed model input, the specific inputs to that Reshape op, or the use of -1 in its target shape attribute, prevents TFLite's static shape inference from resolving the full output shape during delegate preparation. The GPU delegate is stricter than the CPU here.

Q2: **Manual Partitioning?** Possible via C++ modifications or advanced delegate APIs but complex and often leads to performance loss due to CPU and GPU overhead. Numerical consistency differences between CPU/GPU ops are also possible. It's far better to fix the underlying shape issue. `max_delegate_partitions` just limits the number of GPU subgraphs it doesn't selectively choose ops based on shape safety.

Final Q **Making ViT GPU-Compatible**: Ensure conversion with strict static input shapes (no `None`) and use `Netron/benchmark_model` to pinpoint the exact ops causing the dynamic shape issue or analyze why that ops output shape is not static.

Please refer this similar issue https://github.com/tensorflow/tensorflow/issues/48553, If I have missed something here please let me know. Thank you for your cooperation and patience.

**Comment by alstjd025** [comment]
Hi @gaikwadrahul8.
Thank you for the kind response.

You are saying that modifying the source code of TFLite to delegate only specific operators (the way that I mentioned was modifying the variable `nodes_to_replace` at `ModifyGraphWithDelegate() `fuction) is not recommeneded and
the best option to delegate a ViT model with GPU is replacing dynamic shaped operators (e.g., RESHAPE above) with help of Netron visualizer or something.. as i understand.

Here's another questions. :P
In my hands on experiences, most of ViT models (also including LLMs) are not compatible for GPU, especially with large `max_delegate_partitions` parameters. 
1. Is there any plan to make compatible those transformer based models for GPU?
2. Do you have any information about GPU delegate compatible ViT or LLM models?(some unsupported operators are OK but without explicit errors) it would be great if you can offer me.

Thank you!

**Comment by alstjd025** [comment]
I have no idea on "Closed as not planned"... i just misclicked :(

**Comment by svobora** [comment]
This project is dead. It is 2025 and they still dont support acceleration of 4 year old models, let alone have a model zoo. This project needs new people and roadmaps. First thing to do is to prepare model zoo with BASIC model types like MobileNet, ResNet, ViT, and some others, lets say 50 models as competition does, then fix all delegates to enable each network to be converted. I have spoken.

**Comment by gaikwadrahul8** [comment]
Hi, @alstjd025 
Thank you for the kind response and for detailing your experience. Your understanding is correct modifying the TFLite source code to delegate specific operators is not the recommended approach as it's complex and can lead to performance issues due to CPU/GPU overhead. The best strategy remains to address the dynamic shape and operator compatibility problems within the model itself. To answer your questions about future support for transformer-based models:

**1. Is there any plan to make compatible those transformer based models for GPU?**
Yes, The [AI Edge Torch Generative API](https://github.com/google-ai-edge/ai-edge-torch/tree/main/ai_edge_torch/generative/) is a mobile-optimized library for converting transformer-based PyTorch models to the LiteRT format, enabling you to convert Large Language Models (LLMs) for on-device image and text generation with LiteRT and MediaPipe. The AI Edge Torch Generative API is currently CPU-only, with planned GPU and NPU support.

**2. Do you have any information about GPU delegate compatible ViT or LLM models?**
A definitive public list of compatible models is not readily available. Many ViT and LLM models face significant challenges with GPU delegation due to the reasons you've identified such as dynamic shapes and op support issues. The TFLite team provides several examples of models built to take advantage of GPU acceleration such as MobileNet and PoseNet but these are primarily convolutional neural networks optimized for vision tasks rather than transformers. The current recommendation for ViT and LLMs is to rely on the latest conversion and debugging tools like [AI Edge Torch Generative API](https://github.com/google-ai-edge/ai-edge-torch/tree/main/ai_edge_torch/generative/) adapt the models for compatibility. 

Thank you for your cooperation and patience.
